### ISSUE TITLE
JetEngine v2.9.0 Beta

### ISSUE THREAD

**Issue by MjHead** [CrocoBlock/suggestions#3803]
Hi,

We are almost getting ready to release JetEngine v2.9.0 and we would like to ask you to help us test JetEngine v2.9.0 Beta.

You can download beta version from your Crocoblock account.

### Import glossaries from the file

From now on, there’s no need to add all options to the glossary manually. This feature lets you import a file from which these options will be parsed. It is convenient when it comes to large amounts of data. For example, you downloaded a CSV file storing lists of all countries or car brands, and you can upload it as a glossary instead of setting 100+ options by hand. Currently, you can import glossary options from **CSV** and **JSON** files.

The **Data Source** option has been added to the Glossary settings: http://prntscr.com/1jclgmn. If you select the “Get items from uploaded file” option in this drop-down select, new settings will appear below:
- File to Get Data From – a file from the media library from which the items will be imported;
- Value Column – a column (or a key for JSON files) from which the values ​​for glossary options will be taken;
- Label Column – a column (or a key for JSON files) from which the names for options will be taken.

### Calculated Columns and Grouping Results for SQL Queries

These two features work best together. Grouping allows you to combine query results by duplicate values ​​in one of the columns. For example, you have a table column with a list of users, and one of the columns is the user’s country of residence. You can group the selection by this column, and then only one of the users from each country will be displayed by default. In isolation, this feature can rarely be helpful. Still, if you combine it with the Include Calculated Columns feature, you could display a list of countries and some calculated values ​​for all rows that contain any given country. For example, the min, max, average value, amount, counter (for example, the country and the number of users from this country in the database).

The **Grouping feature** is implemented by two options – http://prntscr.com/1g5barz:
- Group Results – it enables the grouping by a specific column; 
- Group By – here, you need to select the column by which to group the results.

The **Include Calculated Columns** feature first needs to be enabled: http://prntscr.com/1g5bth8. Then you add new calculated columns. For each column, it is necessary to select the table’s column based on which the calculations will be carried out and apply the desired function: http://prntscr.com/1g5cibr.  

### Query Results for Dynamic Functions

Before this release, Dynamic Functions worked only with standard WP objects like posts and terms. Now you can display a single result from any custom SQL query in a dynamic function. You can display the value of a column (in this case, if the query returns more than one result, then the value of the selected column from the first obtained result will be displayed). Or, you can apply one of the available functions to the selected column (in this case, if the query returns more than one row, then the result of applying the function to all values ​​of the selected column from all rows will be displayed). The primitive example of using this feature is the calculation of some dynamic values ​​for the CCT.

- http://prntscr.com/1g5qm87 – Select “SQL query results” in dynamic function settings;
- http://prntscr.com/1g5ref6 – Select the query column to be shown in the results;
- http://prntscr.com/1g5rnu9 – Select the function to apply to the results.

### External Modules

As of version 2.9.0, we start dividing JetEngine modules into Default and External. On the client side, this feature means little – it still allows enabling/disabling additional functionality on the modules page, yet the modules themselves are divided into two sections. On the technical side, it means more. Under the hood, external modules become separate plugins, i.e., their code does not depend on the main JetEngine code. What does it bring? It means less code in the primary plugin, making it easier to update and maintain large modules. Since they are not tied to the JetEngine code now, we’ll release updates with bug fixes and new features much faster and safer. Alternatively, if the module’s code is in JetEngine, it can only be updated with the core plugin. Every time we need to update JetEngine, we are supposed to test it thoroughly, not just the module itself. Now, we will be able to deliver external modules separately, which means we should test only the module’s code before releasing it.

A new section has been added to the module management interface – http://prntscr.com/1g34qp0. When activating modules from this section, the corresponding plugin will be installed and activated. The names of the plugins related to this module can be seen in the popup window. The related plugins can be updated here, as well. External modules are available only after license activation.

### Dynamic Tables Module

In the future, all big external modules will be added along with JetEngine major releases since they commonly require changes in the plugin source code. The first such module is Dynamic Tables. All in all, the whole point of the module is in the name. It allows you to create different tables based on dynamic data, be it tables of posts, WooCommerce products, users, terms, comments, etc. You can also build tables based on CCT data or tables with data queried directly from the SQL database. In the table cells, you can either display the data pulled from the database or Listing templates built based on this data. Also, the Dynamic Tables module is compatible with JetSmartFilters, so the selected data can be filtered or sorted as you like.

After you activate the module, a new subitem called Tables Builder will be added to the JetEngine menu – http://prntscr.com/1g3nxlx. Here you can create, edit, and delete tables. The interface is basically the same as for other Crocoblock components. To display the created table on the page, you’ll need either some Elementor widget or the corresponding block in the block editor. Consider this:

- http://prntscr.com/1g3oml9 – Tables are always Query-based. The logic is the same as with the Listing Grid. A Query determines what data will be selected from the database and how, while the table is only responsible for displaying this data;
- http://prntscr.com/1g3p5jl – For easier access to information inside the Query, you can make a test query during which the table determines which columns are inside the query and returns a list of these columns as options to be displayed on the frontend;
- http://prntscr.com/1g3pslp – In the Columns tab, we specifically configure the data that will be shown in the table;
- http://prntscr.com/1g3t054 – In each column, you can display either a raw value or a Listing template;
- http://prntscr.com/1g3te04 – The raw value can be taken from the current object, metadata, or columns obtained from the Query. The raw value can then be processed similarly to the Dynamic Field widget – you can apply a callback to it, change the format, etc.;
- http://prntscr.com/1g3ygl5 – the main styles of the table are configured in the widget that displays it. But in the table itself, you can set up some specific styles that will apply to certain columns only;
- http://prntscr.com/1g43ft6 – when you click on the “Reload Preview” button, you can load a simplified preview of the table. This preview is mainly intended for content preview, not styles, given that table styles in the current theme will be different, and the table can be additionally styled during the output.

### Dynamic Charts Module

This is also an external module. It is designed to display dynamic data in graph format. By definition, this module is very similar to dynamic tables. The only difference is that it shows data on the front-end in the form of a graph.

After activating the module, the Charts Builder subitem is added to the JetEngine menu. You can add, edit, and delete charts here. To display the created chart in the front, need either some Elementor widget or the corresponding block in the block editor. Consider this:

- http://prntscr.com/1g49ely – before starting, you need to select the chart type to be displayed;
- When setting up the columns of the graph, you need to take into account that the first column will be used for labels and the remaining columns will be used for values;
- Data output in columns can be configured in the same way as for tables, except that data output in the format of templates is - not available for graphs since the graph will always require only “raw” values.

#### Notes

For bugs and remarks, please reply with a comment on this issue here on GitHub.

Please keep this related to the Current beta Only! any other issue or feature request, feel free to open a new issue.

Many thanks for your support and help!

**Comment by jamsheehan** [comment]
Getting this when attempting to activate some of the external packages: 

The package could not be installed. PCLZIP_ERR_BAD_FORMAT (-10) : Unable to find End of Central Dir Record signature

https://prnt.sc/1lz0b9q

- NGNIX server (running local copy of a client site)
- WP 5.8
- 256MB RAM 

@MjHead same thing, so I created a new subkey licence and it didn't work. When I used the main account licence it worked first time. 

Also tried it on a test site, same result. So I suspect it can't read the subkey licences. Hope this helps!

**Comment by MjHead** [comment]
@jamsheehan Hi! Could you please try deactivate and than activate your Crocoblock license again?

**Comment by iamdanbarrett** [comment]
In the Charts Builder setup area, beneath the dropdown 'Data Query' it is missing the URL to go create a query, it is there on the Tables builder. It isn't obvious as to what to do either until I saw that link.

<img width="831" alt="Screenshot 2021-08-10 at 20 28 48" src="https://user-images.githubusercontent.com/3041238/128923008-24b9a115-c168-47ff-b0c1-c7002313aa02.png">
<img width="830" alt="Screenshot 2021-08-10 at 20 28 35" src="https://user-images.githubusercontent.com/3041238/128923013-4a7b83a7-686c-419e-8c56-ec6fee6de24f.png">

**Comment by iamdanbarrett** [comment]
**Note!**
We not found any visible columns in the fetched data. You can process build the chart, but `Fetched columns` data source won`t be available for the chart columns

Should be written as...

**Note!**
We have not found any visible columns in the fetched data. You can process & build the chart, but the 'Fetched columns` data source won`t be available for the chart columns.

<img width="581" alt="Screenshot 2021-08-10 at 20 32 19" src="https://user-images.githubusercontent.com/3041238/128923439-1b9987dc-4d41-4710-9046-a4ed23407027.png">

**Comment by jamsheehan** [comment]
**Table Builder**
Not getting an error message if the output is invalid: 

https://prnt.sc/1mc2589

Works fine if you filter the output: 

https://prnt.sc/1mc6291

PHP Fatal error:  Uncaught Error: Object of class stdClass could not be converted to string in /www/wp-content/plugins/jet-engine-dynamic-tables-module/includes/render/table-renderer.php:121"

**Comment by Lonsdale201** [comment]
I played around the Dynamic table. When i selected the data (post/terms/user), and choose the title, and i enabled the filter field output - selected the get post title / linkable. Full broken the page where i used the Dynamic table.  At this page (home page), cant access the elementor, white blank in the frontend. Consolse didnt say anything.

**Comment by Lonsdale201** [comment]
Btw, as always great job, this is an impressive update. :) 

**My question about the dynamic table:**

1, What you think about to select Terms? (similar, like the meta data, but with dropdown with the terms) (like the dynamic terms widget)
2, In a future will be possible to select the post thumbnail directly? ( I know, _thumbnail_id with filter fied output get image by id work like a charm)
 
 Style idea:
 
 Will be in a future add icons directly in the table editor?
 Striped rows background option?(just like the jetelements table) (not a big deal to do this with css i know, but not everybody can do it.)
 
 Thank you for your hard works guys. :)

**Comment by sixmanstone** [comment]
Once again you did a good job.  The Dynamic Tables Module demonstrates this, but I suggest adding JetPopup compatibility for the Dynamic Tables Module so that we can add a popup for Dynamic Field links to see the single post, a Dynamic Fields to edit, and a Dynamic Fields for  delete a publication row in the Dynamic Tables Module.  And why not the possibility of adding in the columns of the table the Dynamic Field widget to see the single publication, modify and delete.  Excuse me for the English I write in French and I translate the text on Google.

**Comment by Lonsdale201** [comment]
> Once again you did a good job. The Dynamic Tables Module demonstrates this, but I suggest adding JetPopup compatibility for the Dynamic Tables Module so that we can add a popup for Dynamic Field links to see the single post, a Dynamic Fields to edit, and a Dynamic Fields for delete a publication row in the Dynamic Tables Module. And why not the possibility of adding in the columns of the table the Dynamic Field widget to see the single publication, modify and delete. Excuse me for the English I write in French and I translate the text on Google.

Jetpopup compability good idea btw, use the [deepl ](https://www.deepl.com/translator) instead of google translater. :)

**Comment by MjHead** [comment]
> **ble Builder**
> Not getting an error message if the output is invalid:
> 
> https://prnt.sc/1mc2589
> 
> Works fine if you filter the output:
> 
> https://prnt.sc/1mc6291
> 
> PHP Fatal error: Uncaught Error: Object of class stdClass could not be converted to string in /www/wp-content/plugins/jet-engine-dynamic-tables-module/includes/render/table-renderer.php:121"

@jamsheehan Hi! Can't reproduce the issue, could you please share a screen of your query config

**Comment by MjHead** [comment]
> I played around the Dynamic table. When i selected the data (post/terms/user), and choose the title, and i enabled the filter field output - selected the get post title / linkable. Full broken the page where i used the Dynamic table. At this page (home page), cant access the elementor, white blank in the frontend. Consolse didnt say anything.

@Lonsdale201 Hi! Can't reproduce the issue, could you please share a screen of your query and table columns config?

**Comment by Lonsdale201** [comment]
> > I played around the Dynamic table. When i selected the data (post/terms/user), and choose the title, and i enabled the filter field output - selected the get post title / linkable. Full broken the page where i used the Dynamic table. At this page (home page), cant access the elementor, white blank in the frontend. Consolse didnt say anything.
> 
> @Lonsdale201 Hi! Can't reproduce the issue, could you please share a screen of your query and table columns config?

Deleted, when shoot down my site :D But i will try to reproduce, and i will make a video record for it. :)

**Comment by P258tiwari** [comment]
(https://user-images.githubusercontent.com/64008801/129224003-5d1c6e7e-8585-410c-bd88-6f0b705f8c4f.png)
Does not active module showing same error

**Comment by P258tiwari** [comment]
![Web capture_13-8-2021_175051_example ampwake com](https://user-images.githubusercontent.com/64008801/129356274-da1cbb12-c6f6-4e7a-8009-27c02705e015.jpeg)


Please check I am not able to access the beta version and this screen shown when I activate the module

**Comment by MjHead** [comment]
@P258tiwari Did you tried deactivate and then activate your Crocoblock license again?

**Comment by P258tiwari** [comment]
> @P258tiwari Did you tried deactivate and then activate your Crocoblock license again?

Yes I tried deactivate and active and subsite license but still same problem

**Comment by sixmanstone** [comment]
Hi, consider adding a footer to the Dynamic Tables Module to insert columns to sum numeric data.

**Comment by Lonsdale201** [comment]
@MjHead Today i tried to reproduce the issue, but did not succeed. :)

**Comment by rousarp** [comment]
I have registered two problems with the forms:

a) form in the popup window is not initialized correctly, preset form values do not work for me - current_id is not transferred (CPT single part template,  button, popup with form, pre-set form field values, source: post, get post id from: current post). If I use JE 2.8.9, the value is transferred.

b) the calculation field does not work dynamically if the form is placed in tabs (inserted using shortcode). Do not change the sum when changing values. AJAX problem? The same form placed under tabs works correctly. But the problem is probably not related to the new version...

Query builder with Custom SQL Query and Tables is great, can the query (and tables) be exported or migrated to another server? Same as CPT?
Thank you for your work,
Pavel

**Comment by pegujurado** [comment]
@MjHead, reading data from current meta user still doesn't work in query builder. This is critical functionality. 
Is there any news? my project is waiting for this functionality and still not seen. th

**Comment by nreljed** [comment]
I will show you a very confusing detail, which has been around for a long time

I created an account page with Profile builder and 4 other pages
dashboard, new post, profile, site settings..

Then I open them in a new tab one by one by

![menu](https://user-images.githubusercontent.com/82481592/130305054-fab8007a-6d40-485d-b243-3744997dbcf0.png)

![home](https://user-images.githubusercontent.com/82481592/130305057-74834e6a-c5dc-4da6-bc0d-df508eb8fa36.png)

Confusion .. no way to understand what is going on .. they are all..

**Account - Home** 

Same thing if I edit 10 posts or 20

I will have 20 Account - Home

instead of for example

Account - post 1
Account - post 2
Account - post 3
Account - Dashboard
Account - My Profile
etc

it's a total confusion :( 

I hope it will finally be corrected

**Comment by nreljed** [comment]
clean install wordpress 
Test slider listing grid 

for some unknown reason jet-slick-dots has 40px padding by default
the dots are not centered

![Capture d’écran 2021-08-24 004427](https://user-images.githubusercontent.com/82481592/130534132-a6cace2b-492f-48f1-b2bc-af44292d953d.png)

**Comment by MjHead** [comment]
> @MjHead, reading data from current meta user still doesn't work in query builder. This is critical functionality.
> Is there any news? my project is waiting for this functionality and still not seen. th

Hi @pegujurado! If you mean using current meta values as dynamic argument for Query, could you please download JetEngine 2.9.0-beta4 from you account and try macros `Current user meta`
![image](https://user-images.githubusercontent.com/4987981/130615331-1ee2b482-bd79-4400-9a59-66da1715fb3d.png)

**Comment by pegujurado** [comment]
> > @MjHead, reading data from current meta user still doesn't work in query builder. This is critical functionality.
> > Is there any news? my project is waiting for this functionality and still not seen. th
> 
> Hi @pegujurado! If you mean using current meta values as dynamic argument for Query, could you please download JetEngine 2.9.0-beta4 from you account and try macros `Current user meta`
> ![image](https://user-images.githubusercontent.com/4987981/130615331-1ee2b482-bd79-4400-9a59-66da1715fb3d.png)

I have no words of thanks! I've already tried it and it works great. Impressive. You are incredible. @MjHead
